offerurl.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+package charm
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/juju/errors"
+	names "gopkg.in/juju/names.v2"
+)
+
+// OfferURL represents the location of an offered application and its
+// associated exported endpoints.
+type OfferURL struct {
+	// Source represents where the offer is hosted.
+	// If empty, the model is another model in the same controller.
+	Source string // "<controller-name>" or "<jaas>" or ""
+
+	// User is the user whose namespace in which the offer is made.
+	// Where a model is specified, the user is the model owner.
+	User string
+
+	// ModelName is the name of the model providing the exported endpoints.
+	// It is only used for local URLs or for specifying models in the same
+	// controller.
+	ModelName string
+
+	// ApplicationName is the name of the application providing the exported endpoints.
+	ApplicationName string
+}
+
+// Path returns the path component of the URL.
+func (u *OfferURL) Path() string {
+	var parts []string
+	if u.User != "" {
+		parts = append(parts, u.User)
+	}
+	if u.ModelName != "" {
+		parts = append(parts, u.ModelName)
+	}
+	path := strings.Join(parts, "/")
+	path = fmt.Sprintf("%s.%s", path, u.ApplicationName)
+	if u.Source == "" {
+		return path
+	}
+	return fmt.Sprintf("%s:%s", u.Source, path)
+}
+
+func (u *OfferURL) String() string {
+	return u.Path()
+}
+
+// AsLocal returns a copy of the URL with an empty (local) source.
+func (u *OfferURL) AsLocal() *OfferURL {
+	localURL := *u
+	localURL.Source = ""
+	return &localURL
+}
+
+// HasEndpoint returns whether this offer URL includes an
+// endpoint name in the application name.
+func (u *OfferURL) HasEndpoint() bool {
+	return strings.Contains(u.ApplicationName, ":")
+}
+
+// modelApplicationRegexp parses urls of the form controller:user/model.application[:relname]
+var modelApplicationRegexp = regexp.MustCompile(`(/?((?P<user>[^/]+)/)?(?P<model>[^.]*)(\.(?P<application>[^:]*(:.*)?))?)?`)
```

Comment:
If you want you can make this stricter by concatenating the  "snippet" regexes from names.v2 (e.g https://github.com/juju/names/blob/v2/application.go#L13). I have used a similar approach for validating the endpoints: https://github.com/juju/charm/blob/v6/bundledata.go#L522-L527